File: services/api/app/seed.py
from datetime import UTC, datetime, timedelta
from decimal import Decimal
import random

from faker import Faker

from app.database import SessionLocal, create_all
from app.models import Customer, Order

fake = Faker("en_IN")
CITIES = ["Hyderabad", "Bengaluru", "Mumbai", "Delhi", "Pune", "Chennai", "Kolkata", "Ahmedabad"]
CATEGORIES = ["Fashion", "Beauty", "Grocery", "Electronics", "Home", "Footwear"]
CHANNELS = ["email", "whatsapp", "sms", "rcs"]


def segment_profile(index: int) -> tuple[int, int]:
    if index < 60:
        return 8, 35
    if index < 170:
        return 2, 25
    if index < 300:
        return 1, 120
    if index < 410:
        return 4, 75
    return 1, 10


def main() -> None:
    create_all()
    db = SessionLocal()
    try:
        if db.query(Customer).count():
            print("Seed skipped: customers already exist.")
            return

        customers: list[Customer] = []
        for index in range(500):
            order_count, max_age = segment_profile(index)
            last_order = datetime.now(UTC) - timedelta(days=random.randint(1, max_age))
            customer = Customer(
                name=fake.name(),
                email=f"customer{index + 1}@example.com",
                phone=fake.phone_number()[:32],
                city=random.choice(CITIES),
                preferred_channel=random.choice(CHANNELS),
                total_spent=Decimal("0"),
                last_order_date=last_order,
                created_at=datetime.now(UTC) - timedelta(days=random.randint(30, 700)),
            )
            db.add(customer)
            db.flush()
            total = Decimal("0")
            for _ in range(order_count):
                amount = Decimal(random.randint(350, 8500))
                total += amount
                db.add(
                    Order(
                        customer_id=customer.id,
                        amount=amount,
                        category=random.choice(CATEGORIES),
                        order_date=last_order - timedelta(days=random.randint(0, 240)),
                    )
                )
            customer.total_spent = total
            customers.append(customer)

        while db.query(Order).count() < 2000:
            customer = random.choice(customers)
            amount = Decimal(random.randint(250, 6500))
            customer.total_spent += amount
            order_date = datetime.now(UTC) - timedelta(days=random.randint(1, 365))
            if not customer.last_order_date or order_date > customer.last_order_date:
                customer.last_order_date = order_date
            db.add(Order(customer_id=customer.id, amount=amount, category=random.choice(CATEGORIES), order_date=order_date))

        db.commit()
        print("Seeded 500 customers and 2000 orders.")
    finally:
        db.close()


if __name__ == "__main__":
    main()

File: services/node-api/src/seed.ts
import { connectDb } from "./db.js";
import { Campaign, Communication, CommunicationEvent, Customer, Order, Segment } from "./models.js";
import { calculateCustomerIntelligence } from "./services/intelligence.js";

const cities = ["Hyderabad", "Bengaluru", "Mumbai", "Delhi", "Pune", "Chennai", "Kolkata", "Ahmedabad"];
const genders = ["Female", "Male", "Non-binary"];
const categories = ["Fashion", "Shoes", "Beauty", "Electronics", "Home", "Grocery"];
const products = ["Sneakers", "Kurta", "Skincare Kit", "Headphones", "Coffee Maker", "Denim Jacket", "Running Shoes"];

function pick<T>(items: T[]) {
  return items[Math.floor(Math.random() * items.length)];
}

function daysAgo(days: number) {
  return new Date(Date.now() - days * 24 * 60 * 60 * 1000);
}

await connectDb();
await Promise.all([Customer.deleteMany({}), Order.deleteMany({}), Segment.deleteMany({}), Campaign.deleteMany({}), Communication.deleteMany({}), CommunicationEvent.deleteMany({})]);

const customers = [];
for (let index = 0; index < 1000; index++) {
  const orderCount = index < 120 ? 12 : index < 400 ? 6 : index < 760 ? 3 : 1;
  const inactive = index % 5 === 0;
  const lastPurchaseDate = daysAgo(inactive ? 90 + Math.floor(Math.random() * 140) : Math.floor(Math.random() * 45) + 1);
  const customer = await Customer.create({
    name: `Customer ${index + 1}`,
    email: `customer${index + 1}@example.com`,
    phone: `+91${9000000000 + index}`,
    city: pick(cities),
    gender: pick(genders),
    age: 18 + Math.floor(Math.random() * 45),
    totalSpend: 0,
    orderCount,
    lastPurchaseDate
  });
  customers.push(customer);
}

let createdOrders = 0;
for (const customer of customers) {
  let totalSpend = 0;
  const initialOrderCount = Number(customer.orderCount ?? 0);
  for (let count = 0; count < initialOrderCount; count++) {
    const amount = 300 + Math.floor(Math.random() * 12000);
    totalSpend += amount;
    await Order.create({
      customerId: customer._id,
      amount,
      product: pick(products),
      category: pick(categories),
      date: daysAgo(Math.floor(Math.random() * 365))
    });
    createdOrders++;
  }
  customer.totalSpend = totalSpend;
  Object.assign(
    customer,
    calculateCustomerIntelligence({
      totalSpend,
      orderCount: initialOrderCount,
      lastPurchaseDate: customer.lastPurchaseDate ?? new Date()
    })
  );
  await customer.save();
}

while (createdOrders < 5000) {
  const customer = pick(customers);
  const amount = 250 + Math.floor(Math.random() * 9000);
  await Order.create({ customerId: customer._id, amount, product: pick(products), category: pick(categories), date: daysAgo(Math.floor(Math.random() * 365)) });
  customer.totalSpend = Number(customer.totalSpend ?? 0) + amount;
  customer.orderCount = Number(customer.orderCount ?? 0) + 1;
  Object.assign(
    customer,
    calculateCustomerIntelligence({
      totalSpend: Number(customer.totalSpend ?? 0),
      orderCount: Number(customer.orderCount ?? 0),
      lastPurchaseDate: customer.lastPurchaseDate ?? new Date()
    })
  );
  await customer.save();
  createdOrders++;
}

const cutoff = daysAgo(90);
const segment = await Segment.create({
  name: "Inactive high-value shoppers",
  rules: { inactiveDays: 90, minSpend: 5000 },
  audienceSize: await Customer.countDocuments({ totalSpend: { $gte: 5000 }, lastPurchaseDate: { $lte: cutoff } })
});

await Campaign.create({
  name: "Winback 90",
  objective: "Bring back customers inactive for 90 days",
  segmentId: segment._id,
  channel: "WhatsApp",
  message: "Hi {{name}}, we missed you. Enjoy 20% off your next order.",
  status: "DRAFT"
});

console.log("Seeded 1000 customers and 5000 orders.");
process.exit(0);
